CMakeLists.txt

```diff
@@ -0,0 +1,397 @@
+    RPC/OpaqueClientRPC.cc
+    RPC/OpaqueServer.cc
+    RPC/OpaqueServerRPC.cc
+    RPC/Protocol.cc
+    RPC/Server.cc
+    RPC/ServerRPC.cc
+    RPC/ThreadDispatchService.cc
+    )
+
+set(EVENT_SRC
+    Event/File.cc
+    Event/Loop.cc
+    Event/Signal.cc
+    Event/Timer.cc
+    )
+
+set(SERVER_SRC
+    Server/ClientService.cc
+    Server/ControlService.cc
+    Server/Globals.cc
+    Server/RaftConsensus.cc
+    Server/RaftConsensusInvariants.cc
+    Server/RaftService.cc
+    Server/ServerStats.cc
+    Server/StateMachine.cc
+    ${SERVER_PROTO_SRC}
+    )
+
+set(STORAGE_SRC
+    Storage/FilesystemUtil.cc
+    Storage/Layout.cc
+    Storage/Log.cc
+    Storage/LogFactory.cc
+    Storage/MemoryLog.cc
+    Storage/SegmentedLog.cc
+    Storage/SimpleFileLog.cc
+    Storage/SnapshotFile.cc
+    ${STORAGE_PROTO_SRC}
+    )
+
+set(TREE_SRC
+    Tree/ProtoBuf.cc
+    Tree/Tree.cc
+    ${TREE_PROTO_SRC}
+    )
+
+add_custom_command(
+    OUTPUT build
+    COMMAND mkdir -p build
+    )
+
+macro(c_proto_buff INPUT_LIST OUTPUT_LIST)
+    list(LENGTH ${INPUT_LIST} LIST_SIZE)
+    math(EXPR MAX_INDEX ${LIST_SIZE}-1)
+
+    ## get the first output to make the output dir
+    list(GET ${OUTPUT_LIST} 0 CUR_OUTPUT)
+
+    ## create dir prefix
+    string(REGEX MATCH build/.*/ OUTPUT_PATH ${CUR_OUTPUT})
+    add_custom_command(
+        OUTPUT ${OUTPUT_PATH}
+        COMMAND mkdir -p ${OUTPUT_PATH}
+        DEPENDS build
+        )
+
+    foreach(INDEX RANGE 0 ${MAX_INDEX}-1)
+        list(GET ${INPUT_LIST} ${INDEX} CUR_INPUT)
+        list(GET ${OUTPUT_LIST} ${INDEX} CUR_OUTPUT)
+
+        add_custom_command(
+            OUTPUT ${CUR_OUTPUT}
+            COMMAND cp ${CUR_INPUT} build/${CUR_INPUT} && protoc -I. --cpp_out=. --python_out=. build/${CUR_INPUT}
+            DEPENDS ${CUR_INPUT} ${OUTPUT_PATH}
+            )
+    endforeach()
+
+endmacro(c_proto_buff)
+
+c_proto_buff(CORE_PROTO CORE_PROTO_SRC)
+c_proto_buff(PROTOCOL_PROTO PROTOCOL_SRC)
+c_proto_buff(TREE_PROTO TREE_PROTO_SRC)
+c_proto_buff(SERVER_PROTO SERVER_PROTO_SRC)
+c_proto_buff(STORAGE_PROTO STORAGE_PROTO_SRC)
+
+##executable daemon
+## I will change it back to LogCabin after I finish all coding
+add_executable(LogCabinServer
+    Server/Main.cc
+    ${SERVER_SRC}
+    ${STORAGE_SRC}
+    ${CLIENT_SRC}
+    ${TREE_SRC}
+    ${PROTOCOL_SRC}
+    ${RPC_SRC}
+    ${EVENT_SRC}
+    ${CORE_SRC}
+    )
+
+add_executable(ServerControl
+    Client/ServerControl.cc
+    )
+
+add_library(logcabin STATIC 
+    EXCLUDE_FROM_ALL
+    ${CLIENT_SRC}
+    ${TREE_SRC}
+    ${PROTOCOL_SRC}
+    ${RPC_SRC}
+    ${EVENT_SRC}
+    ${CORE_SRC}
+    )
+
+add_executable(StorageTool
+    Storage/Tool.cc
+    ${STORAGE_SRC}
+    ${SERVER_PROTO_FOR_STORAGE_TOOL_SRC}
+    ${TREE_SRC}
+    ${PROTOCOL_SRC}
+    ${CORE_SRC}
+    )
+
+add_custom_target(check
+    COMMAND scripts/cpplint.py
+    )
+add_custom_target(lint
+    COMMAND scripts/cpplint.py
+    )
+add_custom_target(docs
+    COMMAND doxygen docs/Doxyfile
```

Comment:
docs/Doxyfile EXCLUDE (around line 800) should exclude CMakeFiles directory too.